### PR TITLE
[Assistant Generation] Improve Encoder Decoder

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR speeds up assistant generation / speculative decoding for encoder-decoder models such as Distill-Whisper by ~20-30%.

Improvements:
- If assistant and model share same encoder, let's allow the user to pass `assistant_encoder_outputs` so that the inputs are not encoded twice (gives ~20% speed-up)
- In the small loop I don't think we have to allocate tensors for the attention mask all the time. This is done automatically by the model if necessary (gives ~3,4% speed-up)
- The heuristic to increase / decrease the number of "look-ahead" tokens doesn't work well for whisper, can we maybe allow the user to somehow disable it? Maybe via a config attribute?

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -4529,10 +4530,10 @@ def assisted_decoding(
             # 6. Adjust the max number of assistant tokens to use in the next iteration. This is a simple heuristic,
             # probably can be improved -- we want to balance the benefits of getting assistant tokens correct with the
             # cost of forecasting incorrect assistant tokens.
-            if n_matches == int(assistant_model.max_assistant_tokens):
-                assistant_model.max_assistant_tokens += 2.0
-            else:
-                assistant_model.max_assistant_tokens = max(1.0, assistant_model.max_assistant_tokens - 1.0)
+            # if n_matches == int(assistant_model.max_assistant_tokens):
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:4533]: The heuristic to increase / decrease the number of "look-ahead" tokens doesn't work well for whisper, can we maybe allow the user to somehow disable it? Maybe via a config attribute?

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -4391,19 +4395,16 @@ def assisted_decoding(
                     # `new_token_len` can be 1 or 2 (next token in assistant + last token picked by the larger model)
                     new_token_len = candidate_input_ids.shape[1] - prev_seq_len
                     assist_inputs = candidate_input_ids[:, -new_token_len:]
-                    assist_attn = torch.ones_like(candidate_input_ids)
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:4394]: Do we really need this @gante ? Allocating new memory here every time leads to some slow downs that are not insignificant for Distil Whisper

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -4484,18 +4485,18 @@ def assisted_decoding(
             # 2.2. Process the new logits
             new_logits = outputs.logits[:, -candidate_length - 1 :]  # excludes the input prompt if present
             if len(logits_processor) > 0:
-                for i in range(candidate_length):
+                for i in range(candidate_length + 1):
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:4488]: That was a bug previously. We forgot to apply the logits processors to the last logit here

File: src/transformers/generation/utils.py [hunk 4]
```diff
@@ -4484,18 +4485,18 @@ def assisted_decoding(
...
                     new_logits[:, i, :] = logits_warper(candidate_input_ids[:, : cur_len + i], new_logits[:, i, :])
 
             # 3. Obtain the next tokens from the original model logits.
             if do_sample:
-                probs = new_logits[:, -candidate_length - 1 :, :].softmax(dim=-1)
+                probs = new_logits.softmax(dim=-1)
                 selected_tokens = torch.multinomial(probs[0, :, :], num_samples=1).squeeze(1)[None, :]
             else:
-                selected_tokens = new_logits[:, -candidate_length - 1 :, :].argmax(dim=-1)
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:4498]: This is unnecessary here as `new_logits` is already sliced

File: src/transformers/generation/configuration_utils.py [hunk 5]
```diff
@@ -227,6 +227,20 @@ class GenerationConfig(PushToHubMixin):
         decoder_start_token_id (`int`, *optional*):
             If an encoder-decoder model starts decoding with a different token than *bos*, the id of that token.
 
+        > Generation parameters exclusive to [assistant generation](https://arxiv.org/abs/2211.17192)
+
+        max_assistant_tokens (`int`, *optional*, defaults to 5):
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:232]: Perhaps we can take the chance to give a better name to this variable: `assistant_tokens` or similar. `max_assistant_tokens` implies that the assistant will never cross this limit but, as we can see in `max_assistant_tokens_schedule` (which should also be renamed accordingly), that is not true :)

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -4310,8 +4314,14 @@ def assisted_decoding(
         ["It might be possible to get a better understanding of the nature of the problem, but it's not"]
         ```"""
         # Assistant: initialize assistant-related variables
-        if not hasattr(assistant_model, "max_assistant_tokens"):
-            assistant_model.max_assistant_tokens = 5  # this value, which will be updated, persists across calls
+        if hasattr(assistant_model, "max_assistant_tokens"):
+            warnings.warn(
+                "Setting `max_assistant_tokens` via `assistant_model.max_assistant_tokens` is deprecated and will be removed in v5. Make sure to set `max_assistant_tokens` via the generation_config instead.",
```
Comment (gante) [inline src/transformers/generation/utils.py:4319]: Perhaps we can deprecate this earlier (like in v4.37)?

I haven't seen users fiddling with this internal variable :)

File: src/transformers/generation/utils.py (hunk 2 above)
Comment (gante) [reply src/transformers/generation/utils.py:4394]: Yes, this makes sense to remove, since it is the default attention mask! 👍

File: src/transformers/generation/utils.py (hunk 3 above)
Comment (gante) [reply src/transformers/generation/utils.py:4494]: Good catch 👀

**Review by gante — APPROVED** [review]
Fantastic, thank you for the upgrades @patrickvonplaten 🔥 

Only added two minor, optional nits.

File: src/transformers/generation/configuration_utils.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/generation/configuration_utils.py:232]: Sounds good!

File: src/transformers/generation/utils.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/transformers/generation/utils.py:4319]: Ok for me!

File: examples/research_projects/jax-projects/big_bird/bigbird_flax.py [hunk 7]
```diff
@@ -9,13 +9,13 @@
...
 import joblib
 import optax
-import wandb
 from flax import jax_utils, struct, traverse_util
 from flax.serialization import from_bytes, to_bytes
 from flax.training import train_state
 from flax.training.common_utils import shard
 from tqdm.auto import tqdm
 
+import wandb
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/big_bird/bigbird_flax.py:18]:
```suggestion
```

File: examples/research_projects/jax-projects/big_bird/bigbird_flax.py [hunk 8]
```diff
@@ -9,13 +9,13 @@
 import jax.numpy as jnp
 import joblib
 import optax
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/big_bird/bigbird_flax.py:11]:
```suggestion
import optax
import wandb
```

File: examples/research_projects/jax-projects/big_bird/train.py [hunk 9]
```diff
@@ -2,11 +2,11 @@
 from dataclasses import replace
 
 import jax
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/big_bird/train.py:4]:
```suggestion
import jax
import wandb
```

File: examples/research_projects/jax-projects/big_bird/train.py [hunk 10]
```diff
@@ -2,11 +2,11 @@
...
-import wandb
 from bigbird_flax import Args, DataCollator, FlaxBigBirdForNaturalQuestions, Trainer, build_tx, train_step, val_step
 from datasets import load_dataset
 from flax import jax_utils
 
+import wandb
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/big_bird/train.py:9]:
```suggestion
```

File: examples/research_projects/vqgan-clip/VQGAN_CLIP.py [hunk 11]
```diff
@@ -4,12 +4,12 @@
 import imageio
 import torch
 import torchvision
```
Comment (patrickvonplaten) [inline examples/research_projects/vqgan-clip/VQGAN_CLIP.py:6]:
```suggestion
import torchvision
import wandb
```

File: examples/research_projects/vqgan-clip/VQGAN_CLIP.py [hunk 12]
```diff
@@ -4,12 +4,12 @@
...
-import wandb
 from img_processing import custom_to_pil, loop_post_process, preprocess, preprocess_vqgan
 from loaders import load_vqgan
 from PIL import Image
 from torch import nn
 
+import wandb
```
Comment (patrickvonplaten) [inline examples/research_projects/vqgan-clip/VQGAN_CLIP.py:12]:
```suggestion
```

File: src/transformers/models/biogpt/modeling_biogpt.py [hunk 13]
```diff
@@ -544,7 +544,11 @@ def forward(
             inputs_embeds = self.embed_tokens(input) * self.embed_scale
 
         if attention_mask is None:
-            attention_mask = torch.ones(inputs_embeds.shape[:2], dtype=torch.bool, device=inputs_embeds.device)
```
Comment (patrickvonplaten) [inline src/transformers/models/biogpt/modeling_biogpt.py:547]: This was a bug previously. The attention_mask should be equal to input_embeds + past_key_values length.

**Review by ArthurZucker — APPROVED** [review]
Very clean thanks for improving the performances!
